cpp/src/barretenberg/stdlib/primitives/safe_uint/safe_uint.hpp

```diff
@@ -102,25 +102,32 @@ template <typename ComposerContext> class safe_uint_t {
                          std::string const& description = "") const
     {
         ASSERT(difference_bit_size <= MAX_BIT_NUM);
-        ASSERT(!(this->value.is_constant() && other.value.is_constant()));
+        if constexpr (!IsSimulator<ComposerContext>) {
+            ASSERT(!(this->value.is_constant() && other.value.is_constant()));
+        }
         field_ct difference_val = this->value - other.value;
         safe_uint_t<ComposerContext> difference(difference_val, difference_bit_size, format("subtract: ", description));
         // This checks the subtraction is correct for integers without any wraps
-        if (difference.current_max + other.current_max > MAX_VALUE)
+        if (difference.current_max + other.current_max > MAX_VALUE) {
             throw_or_abort("maximum value exceeded in safe_uint subtract");
+        }
         return difference;
     }
 
     safe_uint_t operator-(const safe_uint_t& other) const
     {
+        if constexpr (!IsSimulator<ComposerContext>) {
+            // WORKTODO: is this really the right condition?
```

Comment:
TODO: I didn't check that the mixed case (one witness one constant) is handled correctly. I wondered if the `&&`s should be `||`s.